### ISSUE TITLE
Test cluster's registry-cred pod stuck in PodInitializing state

### ISSUE THREAD

**Issue by myechuri** [elotl/kip#66]
Provisioned test cluster using https://github.com/elotl/kip/tree/master/deploy/terraform . The cluster came up fine, but `registry-cred` pod is stuck in `PodInitializing` state. 

```
$ kubectl cluster-info
Kubernetes master is running at https://10.0.17.195:6443
KubeDNS is running at https://10.0.17.195:6443/api/v1/namespaces/kube-system/services/kube-dns:dns/proxy

To further debug and diagnose cluster problems, use 'kubectl cluster-info dump'.
```

```
$ kubectl get nodes
NAME                          STATUS   ROLES    AGE   VERSION
ip-10-0-17-195.ec2.internal   Ready    master   24m   v1.18.1
virtual-kubelet               Ready    agent    23m   
```

```
$ kubectl -n kube-system get pods
NAME                                                  READY   STATUS            RESTARTS   AGE
aws-node-tq9z6                                        1/1     Running           0          21m
coredns-66bff467f8-6ml6c                              1/1     Running           0          21m
coredns-66bff467f8-xr29g                              1/1     Running           0          21m
etcd-ip-10-0-17-195.ec2.internal                      1/1     Running           0          21m
kube-apiserver-ip-10-0-17-195.ec2.internal            1/1     Running           0          21m
kube-controller-manager-ip-10-0-17-195.ec2.internal   1/1     Running           0          21m
kube-proxy-pclwn                                      1/1     Running           0          21m
kube-scheduler-ip-10-0-17-195.ec2.internal            1/1     Running           0          21m
registry-creds-l5kzs                                  0/1     PodInitializing   0          21m
virtual-kubelet-db6f48888-dk74s                       1/1     Running           0          21m
```

```
$ kubectl -n kube-system logs registry-creds-l5kzs 
ERROR: logging before flag.Parse: E0415 07:43:53.734772       8 reflector.go:199] github.com/upmc-enterprises/registry-creds/vendor/k8s.io/client-go/tools/cache/reflector.go:94: Failed to list *v1.Namespace: namespaces is forbidden: User "system:serviceaccount:kube-system:default" cannot list resource "namespaces" in API group "" at the cluster scope
ERROR: logging before flag.Parse: E0415 07:43:54.736143       8 reflector.go:199] github.com/upmc-enterprises/registry-creds/vendor/k8s.io/client-go/tools/cache/reflector.go:94: Failed to list *v1.Namespace: namespaces is forbidden: User "system:serviceaccount:kube-system:default" cannot list resource "namespaces" in API group "" at the cluster scope
ERROR: logging before flag.Parse: E0415 07:43:55.737450       8 reflector.go:199] github.com/upmc-enterprises/registry-creds/vendor/k8s.io/client-go/tools/cache/reflector.go:94: Failed to list *v1.Namespace: namespaces is forbidden: User "system:serviceaccount:kube-system:default" cannot list resource "namespaces" in API group "" at the cluster scope
ERROR: logging before flag.Parse: E0415 07:43:56.738964       8 reflector.go:199] github.com/upmc-enterprises/registry-creds/vendor/k8s.io/client-go/tools/cache/reflector.go:94: Failed to list *v1.Namespace: namespaces is forbidden: User "system:serviceaccount:kube-system:default" cannot list resource "namespaces" in API group "" at the cluster scope
ERROR: logging before flag.Parse: E0415 07:43:57.740308       8 reflector.go:199] github.com/upmc-enterprises/registry-creds/vendor/k8s.io/client-go/tools/cache/reflector.go:94: Failed to list *v1.Namespace: namespaces is forbidden: User "system:serviceaccount:kube-system:default" cannot list resource "namespaces" in API group "" at the cluster scope
ERROR: logging before flag.Parse: E0415 07:43:58.741763       8 reflector.go:199] github.com/upmc-enterprises/registry-creds/vendor/k8s.io/client-go/tools/cache/reflector.go:94: Failed to list *v1.Namespace: namespaces is forbidden: User "system:serviceaccount:kube-system:default" cannot list resource "namespaces" in API group "" at the cluster scope
ERROR: logging before flag.Parse: E0415 07:43:59.746607       8 reflector.go:199] github.com/upmc-enterprises/registry-creds/vendor/k8s.io/client-go/tools/cache/reflector.go:94: Failed to list *v1.Namespace: namespaces is forbidden: User "system:serviceaccount:kube-system:default" cannot list resource "namespaces" in API group "" at the cluster scope
ERROR: logging before flag.Parse: E0415 07:44:00.748238       8 reflector.go:199] github.com/upmc-enterprises/registry-creds/vendor/k8s.io/client-go/tools/cache/reflector.go:94: Failed to list *v1.Namespace: namespaces is forbidden: User "system:serviceaccount:kube-system:default" cannot list resource "namespaces" in API group "" at the cluster scope
ERROR: logging before flag.Parse: E0415 07:44:01.749908       8 reflector.go:199] github.com/upmc-enterprises/registry-creds/vendor/k8s.io/client-go/tools/cache/reflector.go:94: Failed to list *v1.Namespace: namespaces is forbidden: User "system:serviceaccount:kube-system:default" cannot list resource "namespaces" in API group "" at the cluster scope
ERROR: logging before flag.Parse: E0415 07:44:02.751224       8 reflector.go:199] github.com/upmc-enterprises/registry-creds/vendor/k8s.io/client-go/tools/cache/reflector.go:94: Failed to list *v1.Namespace: namespaces is forbidden: User "system:serviceaccount:kube-system:default" cannot list resource "namespaces" in API group "" at the cluster scope
```

**Comment by ldx** [comment]
I'm not sure that error in the log is the problem. I provisioned a new cluster:

```ubuntu@ip-10-0-21-170:~$ kubectl get pods -A
NAMESPACE     NAME                                                  READY   STATUS    RESTARTS   AGE
kube-system   aws-node-x2sfb                                        1/1     Running   0          12m
kube-system   coredns-66bff467f8-jpbl7                              1/1     Running   0          12m
kube-system   coredns-66bff467f8-qjrgd                              1/1     Running   0          12m
kube-system   etcd-ip-10-0-21-170.ec2.internal                      1/1     Running   0          13m
kube-system   kube-apiserver-ip-10-0-21-170.ec2.internal            1/1     Running   0          13m
kube-system   kube-controller-manager-ip-10-0-21-170.ec2.internal   1/1     Running   0          13m
kube-system   kube-proxy-sc4fl                                      1/1     Running   0          12m
kube-system   kube-scheduler-ip-10-0-21-170.ec2.internal            1/1     Running   0          13m
kube-system   registry-creds-vjrfp                                  1/1     Running   0          12m
kube-system   virtual-kubelet-db6f48888-d62jg                       1/1     Running   0          12m
```

but I'm getting the same error:

```
ERROR: logging before flag.Parse: E0416 15:58:07.336641       8 reflector.go:199] github.co
m/upmc-enterprises/registry-creds/vendor/k8s.io/client-go/tools/cache/reflector.go:94: Fail
ed to list *v1.Namespace: namespaces is forbidden: User "system:serviceaccount:kube-system:
default" cannot list resource "namespaces" in API group "" at the cluster scope
ERROR: logging before flag.Parse: E0416 15:58:08.338014       8 reflector.go:199] github.co
m/upmc-enterprises/registry-creds/vendor/k8s.io/client-go/tools/cache/reflector.go:94: Fail
ed to list *v1.Namespace: namespaces is forbidden: User "system:serviceaccount:kube-system:
default" cannot list resource "namespaces" in API group "" at the cluster scope
ERROR: logging before flag.Parse: E0416 15:58:09.339379       8 reflector.go:199] github.co
m/upmc-enterprises/registry-creds/vendor/k8s.io/client-go/tools/cache/reflector.go:94: Fail
ed to list *v1.Namespace: namespaces is forbidden: User "system:serviceaccount:kube-system:
default" cannot list resource "namespaces" in API group "" at the cluster scope
```

It'd be interesting to see what kip thinks about the pod state, and kip logs to see the state changes.

**Comment by myechuri** [comment]
Thanks, @ldx ! KIP + vk logs for the pod:

```
I0415 07:30:32.624044       1 opencensus.go:138] sync handled key=kube-system/registry-creds-l5kzs
I0415 07:30:32.624289       1 server.go:643] GetPod "registry-creds-l5kzs"
I0415 07:30:32.624392       1 server.go:575] CreatePod "registry-creds-l5kzs"
I0415 07:30:32.628343       1 opencensus.go:154] Created pod in provider pod=registry-creds-l5kzs namespace=kube-system
I0415 07:30:32.633607       1 opencensus.go:138] processing pod status update key=kube-system/registry-creds-l5kzs
I0415 07:30:32.649148       1 opencensus.go:138] sync handled key=kube-system/registry-creds-l5kzs
I0415 07:30:32.665101       1 server.go:643] GetPod "registry-creds-l5kzs"
I0415 07:30:32.666168       1 opencensus.go:143] Pod registry-creds-l5kzs exists, updating pod in provider pod=registry-creds-
l5kzs namespace=kube-system
I0415 07:30:32.666301       1 server.go:595] UpdatePod "registry-creds-l5kzs"
I0415 07:30:32.666959       1 opencensus.go:154] Updated pod in provider pod=registry-creds-l5kzs namespace=kube-system
I0415 07:30:32.672160       1 opencensus.go:138] processing pod status update key=kube-system/registry-creds-l5kzs
I0415 07:30:38.337962       1 instances.go:270] Starting instance for node: &{{Node v1} {a52d77a8-ac56-45dc-881d-541c883f2f77 
map[] 2020-04-15 07:30:38.336065884 +0000 UTC <nil> map[] 66c4afe5-b2f8-41b1-a625-5e2f9874d216 default} {t3.nano ami-026414670
1c835631 false false {   20G false <nil> false <nil>} {}} {Creating  [] kube-system_registry-creds-l5kzs}}
I0415 07:31:21.434012       1 pod_controller.go:392] Dispatching pod kube-system_registry-creds-l5kzs to node a52d77a8-ac56-45
dc-881d-541c883f2f77
I0415 07:31:21.434804       1 itzoclient.go:381] deploying package to https://10.0.18.102:6421/rest/v1/deploy/kube-system_regi
stry-creds-l5kzs/default-token-wz6ft
I0415 07:31:21.435074       1 server.go:490] kip pod "kube-system_registry-creds-l5kzs" (Dispatching) event {pod-updated pod-r
egistry Pod status phase: 'Dispatching': scheduling pod kube-system_registry-creds-l5kzs onto a52d77a8-ac56-45dc-881d-541c883f
2f77 0xc0009eb900}
I0415 07:31:21.438459       1 itzoclient.go:381] deploying package to https://10.0.18.102:6421/rest/v1/deploy/kube-system_regi
stry-creds-l5kzs/resolvconf
I0415 07:31:21.439538       1 itzoclient.go:381] deploying package to https://10.0.18.102:6421/rest/v1/deploy/kube-system_regi
stry-creds-l5kzs/etchosts
I0415 07:31:21.440394       1 opencensus.go:138] processing pod status update key=kube-system/registry-creds-l5kzs
I0415 07:31:21.441268       1 itzoclient.go:381] deploying package to https://10.0.18.102:6421/rest/v1/deploy/kube-system_regi
stry-creds-l5kzs/resolvconf
I0415 07:31:21.447383       1 itzoclient.go:381] deploying package to https://10.0.18.102:6421/rest/v1/deploy/kube-system_regi
stry-creds-l5kzs/etchosts
I0415 07:31:21.459213       1 itzoclient.go:381] deploying package to https://10.0.18.102:6421/rest/v1/deploy/kube-system_regi
stry-creds-l5kzs/kubeconfig
I0415 07:31:21.469571       1 opencensus.go:138] sync handled key=kube-system/registry-creds-l5kzs
I0415 07:31:21.779447       1 server.go:490] kip pod "kube-system_registry-creds-l5kzs" (Running) event {pod-running pod-contr
oller pod kube-system_registry-creds-l5kzs running on node a52d77a8-ac56-45dc-881d-541c883f2f77 0xc0003dca00}
I0415 07:31:21.806609       1 server.go:643] GetPod "registry-creds-l5kzs"
I0415 07:31:21.808234       1 opencensus.go:143] Pod registry-creds-l5kzs exists, updating pod in provider pod=registry-creds-
l5kzs namespace=kube-system
I0415 07:31:21.813363       1 server.go:490] kip pod "kube-system_registry-creds-l5kzs" (Running) event {pod-updated pod-regis
try Pod status phase: 'Running': Pod is running 0xc000146f00}
I0415 07:31:21.815910       1 server.go:468] no host port mappings found for pod "registry-creds-l5kzs"
I0415 07:31:21.816650       1 opencensus.go:138] processing pod status update key=kube-system/registry-creds-l5kzs
I0415 07:31:21.816109       1 server.go:595] UpdatePod "registry-creds-l5kzs"
I0415 07:31:21.817496       1 opencensus.go:154] Updated pod in provider namespace=kube-system pod=registry-creds-l5kzs
I0415 07:31:21.842336       1 opencensus.go:138] processing pod status update key=kube-system/registry-creds-l5kzs
I0415 07:31:21.852890       1 opencensus.go:138] sync handled key=kube-system/registry-creds-l5kzs
I0415 07:31:22.058123       1 server.go:643] GetPod "registry-creds-l5kzs"
I0415 07:31:22.059355       1 opencensus.go:143] Pod registry-creds-l5kzs exists, updating pod in provider namespace=kube-syst
em pod=registry-creds-l5kzs
I0415 07:31:22.059417       1 server.go:595] UpdatePod "registry-creds-l5kzs"
I0415 07:31:22.059968       1 opencensus.go:154] Updated pod in provider pod=registry-creds-l5kzs namespace=kube-system
I0415 07:31:22.060202       1 opencensus.go:138] sync handled key=kube-system/registry-creds-l5kzs
I0415 07:31:22.065216       1 opencensus.go:138] processing pod status update key=kube-system/registry-creds-l5kzs
I0415 07:31:22.173685       1 server.go:643] GetPod "registry-creds-l5kzs"
I0415 07:31:22.175064       1 opencensus.go:143] Pod registry-creds-l5kzs exists, updating pod in provider namespace=kube-syst
em pod=registry-creds-l5kzs
I0415 07:31:22.175366       1 server.go:595] UpdatePod "registry-creds-l5kzs"
I0415 07:31:22.176164       1 opencensus.go:154] Updated pod in provider namespace=kube-system pod=registry-creds-l5kzs
I0415 07:31:22.186473       1 opencensus.go:138] processing pod status update key=kube-system/registry-creds-l5kzs
I0415 07:31:22.189907       1 opencensus.go:138] sync handled key=kube-system/registry-creds-l5kzs
I0415 07:31:22.301968       1 server.go:643] GetPod "registry-creds-l5kzs"
I0415 07:31:22.303024       1 opencensus.go:143] Pod registry-creds-l5kzs exists, updating pod in provider pod=registry-creds-
l5kzs namespace=kube-system
I0415 07:31:22.303097       1 server.go:595] UpdatePod "registry-creds-l5kzs"
I0415 07:31:22.303646       1 opencensus.go:154] Updated pod in provider pod=registry-creds-l5kzs namespace=kube-system
I0415 07:31:22.303859       1 opencensus.go:138] sync handled key=kube-system/registry-creds-l5kzs
I0415 07:31:22.308884       1 opencensus.go:138] processing pod status update key=kube-system/registry-creds-l5kzs
I0415 07:31:22.412543       1 server.go:643] GetPod "registry-creds-l5kzs"
I0415 07:31:22.415484       1 opencensus.go:143] Pod registry-creds-l5kzs exists, updating pod in provider namespace=kube-syst
em pod=registry-creds-l5kzs
```

**Comment by ldx** [comment]
@justnoise any idea? Based on the logs, the state transition in Kip looks correct.

**Comment by ldx** [comment]
Oh, another problem might be if the pod has probes configured and they fail for some reason. Getting the full pod status via `get pod <registry-pod-...> -oyaml` could be used to confirm this.

**Comment by ldx** [comment]
I think this was due to either the node start race condition fixed in 764384af, or the pod status update fix in 283e979. Both are now in master.
